File: topic1/problem14/Solution.java
class Solution {
    public int canCompleteCircuit(int[] gas, int[] cost) {
        int total = 0, fuel = 0, start = 0;
        for (int i = 0; i < gas.length; i++) {
            total += gas[i] - cost[i];
            fuel += gas[i] - cost[i];

            if (fuel < 0) {
                fuel = 0;
                start = i + 1;
            }
        }
        if (total < 0) return -1;
        return start;
    }
}

File: topic1/problem14/Solution.py
class Solution:
    def canCompleteCircuit(self, gas: List[int], cost: List[int]) -> int:
        if sum(gas) < sum(cost): return -1

        fuel = start = 0
        for i in range(len(gas)):
            fuel += gas[i] - cost[i]

            if fuel < 0:
                fuel = 0
                start = i + 1
        
        return start
